Use rpmcanon in qvm-template

DemiMarie opened the pull request:
This is needed to avoid a security regression in template installation from R4.0.  It is separate from #196 as that commit can be merged sooner.  It requires versions greater than 0.2.2 of qubes-rpm-oxide to be installed in any qube using `qvm-template`.

DemiMarie committed:
Use rpmcanon to canonicalize downloaded templates

rpmcanon is designed to handle potentially malicious packages safely,
and has much less attack surface than RPM itself.  It also prevents
TOCTOU attacks by using O_TMPFILE.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -562,25 +563,32 @@ def qrexec_download(
-            raise ConnectionError(
-                "qrexec call 'qubes.TemplateDownload' failed.")
-
+        read_fd, write_fd = os.pipe()
```
more pythonic way of this is:
```
with subproces.Popen(..., stdin=subprocess.PIPE) as rpmcanon:
     subprocess.Popen(..., stdout=rpmcanon.stdin)
```

DemiMarie committed:
nogpgcheck should be a keyword-only argument

In Python, nonempty strings are truthy, so it was very easy to
accidentally disable signature checking when setting the expected
package name was intended.  Make nogpgcheck a keyword-only argument to
prevent this.

DemiMarie committed:
Strengthen handling of untrusted qrexec output

- Only read 1MiB of stdout and 1KiB of stderr
- Use stricter regular expressions

marmarek commented:
Pylint complains (most issues are about missing parts of rpm python module mockup, but there are other issues too). 
You may try to install `python3-rpm` (see `.gitlab-ci.yml`) - instead of mockup in `test-packages/rpm.py`. I think there was some reason why it wasn't done before, but that was before migrating to gitlab-ci and there is a chance it doesn't apply anymore.

Unit tests fails too, because of missing rpmcanon. This one is a bit tricky, because tests here are running in a generic (fedora?) container, without qubes repos. For now I'd skip those tests in CI, for example like this: `@unittest.skipUnless(spawn.find_executable('rpmcanon'), "rpmcanon not installed")`. BTW, do not hardcode `/usr/bin` path.

marmarek commented on rpm_spec/qubes-core-admin-client.spec.in:
```diff
@@ -17,6 +17,7 @@ BuildRequires:	python%{python3_pkgversion}-xcffib
 Requires:   python%{python3_pkgversion}-qubesadmin
 Requires:   python%{python3_pkgversion}-yaml
 Requires:   qubes-repo-templates
+Requires:   qubes-rpm-oxide > 0.2.2
```
Generally, better use `>= (version that is new enough)` than `> (version that is too old)`. Same in `debian/control`.
But for the purpose of running tests, please move it now to the PR description and add as a commit only when such version will actually exists (I can commit this change too).

marmarek replied:
To clarify: keep (adding) the dependency, just drop the version.

marmarek commented:
Tests failed because the default branch in `qubes-rpm-oxide` is called `main` not `master`, and the PR-building script mishandled that :(
I've started another iteration...
